keps/sig-api-machinery/20190607-owner-ref.md

```diff
@@ -0,0 +1,128 @@
+no lookup or approximation of user intent will be required.  If not provided, the `kind` will be used as it is today.
+If both are provided, old clusters continue to work and new clusters will be more efficient and resilient.
+ 
+```yaml
+ownerReferences:
+- apiVersion: apps/v1
+  kind: DaemonSet
+  resource: daemonsets
+```
+
+## Motivation
+
+* Allow precision.  `kind` is a one to many relationship that makes the exact object reference inefficient or impossible to determine.
+* Remove remote calls.  Because a `kind` to `resource` mapping requires discovery, it is vulnerable to network segmentation
+ 
+### Goals
+
+* Improve reliability. 
+* Improve efficiency.
+* Improve predictability.
+* Avoid disrupting existing clients.
+
+### Non-Goals
+
+* Improving discovery.  No amount of improving speed can fix predictability and we're already doing what caching is practical.
+* Improve rest mapping.  No amount of feature enhancement can fix predictability.
+
+## Proposal
+
+Add a new, optional field to `OwnerReferences` that represents the `resource`.
+
+```go
+type OwnerReference struct {
+	// Kind of the referent.
+	Kind string `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+	// Resource of the referent.
+	// +optional
+	Resource string `json:"resource,omitempty" protobuf:"bytes,8,opt,name=resource"`
+}
+```
+
+The only impact is in the case of a new cluster with both fields set.  In this case, the new cluster does not have to 
+perform any lookups and will trust the client's specification of `resource`.
+
+Existing clients are not required to make this change, but it will improve reliability.  If a client sets the value improperly,
+then the behavior is congruent to behavior with a mis-set `kind`.  If the kind (resource) is valid, but incorrect 
+GC collects the object immediately.  If the kind (resource) is invalid, GC reports the error and waits.  
+
+
+### Behavior of new clusters
+1. `resource` set and `kind` both set: only `resource` is honored and no mapping required  
+2. `resource` set and `kind` not set: validation failure.  Allowing this would break old clusters and clients.  
+3. `resource` not set and `kind` set: `kind` is honored, exactly as before
+4. `resource` not set and `kind` not set: validation failure
+  
+### Behavior of old clusters
+1. `resource` set and `kind` both set: only `kind` is honored, exactly as before  
+2. `resource` set and `kind` not set: validation failure  
+3. `resource` not set and `kind` set: `kind` is honored, exactly as before
+4. `resource` not set and `kind` not set: validation failure
+
+## References
+
+* [OwnerReference struct](https://github.com/kubernetes/apimachinery/blob/kubernetes-1.14.4-beta.0/pkg/apis/meta/v1/types.go#L303-L329)
+* [Example mapping in admission](https://github.com/kubernetes/kubernetes/blob/v1.14.4-beta.0/plugin/pkg/admission/gc/gc_admission.go#L184-L187)
+
+### Test Plan
+
+**blockers for GA:**
+
+* kube-controller-manager should set this field.
```

Comment:
Is this saying kube-controller-manager should set it for all the Kinds it controls, when creating/updating the ownerRefernce?  Or that it should scan all objects in the cluster and append  .reference when it is missing?

Reply:
> Is this saying kube-controller-manager should set it for all the Kinds it controls, when creating/updating the ownerRefernce? Or that it should scan all objects in the cluster and append .reference when it is missing?

When it creates references.  I would not attempt to modify existing ones in the cluster because this KEP isn't attempting to force people to a new API. It is offering a new, more precise API to be used in favor of the existing one by knowledgeable clients.

Going deep in the code, I plan to update `k8s.io/apimachinery/pkg/apis/meta/v1/controller_ref` to add a `NewExactControllerRef` function, leave the old in place, and update the callsites in kube-controller-manager.